source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1119,42 +1171,219 @@ void ClusterManagerImpl::postThreadLocalClusterUpdate(ClusterManagerCluster& cm_
-                        add_or_update_cluster, load_balancer_factory, map = std::move(host_map)](
-                           OptRef<ThreadLocalClusterManagerImpl> cluster_manager) {
-    ThreadLocalClusterManagerImpl::ClusterEntry* new_cluster = nullptr;
-    if (add_or_update_cluster) {
-      if (cluster_manager->thread_local_clusters_.contains(info->name())) {
-        ENVOY_LOG(debug, "updating TLS cluster {}", info->name());
-      } else {
-        ENVOY_LOG(debug, "adding TLS cluster {}", info->name());
-      }
 
-      new_cluster = new ThreadLocalClusterManagerImpl::ClusterEntry(*cluster_manager, info,
-                                                                    load_balancer_factory);
-      cluster_manager->thread_local_clusters_[info->name()].reset(new_cluster);
-    }
+  // Populate the cluster initialization object based on this update.
+  ClusterInitializationObjectConstSharedPtr cluster_initialization_object =
+      addOrUpdateClusterInitializationObjectIfSupported(params, cm_cluster.cluster().info(),
+                                                        load_balancer_factory, host_map);
 
-    for (const auto& per_priority : params.per_priority_update_params_) {
-      cluster_manager->updateClusterMembership(
-          info->name(), per_priority.priority_, per_priority.update_hosts_params_,
-          per_priority.locality_weights_, per_priority.hosts_added_, per_priority.hosts_removed_,
-          per_priority.weighted_priority_health_, per_priority.overprovisioning_factor_, map);
-    }
-
-    if (new_cluster != nullptr) {
+  tls_.runOnAllThreads([info = cm_cluster.cluster().info(), params = std::move(params),
+                        add_or_update_cluster, load_balancer_factory, map = std::move(host_map),
+                        cluster_initialization_object = std::move(cluster_initialization_object)](
+                           OptRef<ThreadLocalClusterManagerImpl> cluster_manager) {
+    // Cluster Manager here provided by the particular thread, it will provide
+    // this allowing to make the relevant change.
+    if (const bool defer_unused_clusters =
+            cluster_initialization_object != nullptr &&
+            !cluster_manager->thread_local_clusters_.contains(info->name()) &&
+            !Envoy::Thread::MainThread::isMainThread();
+        defer_unused_clusters) {
+      // Save the cluster initialization object.
+      ENVOY_LOG(debug, "Deferring add or update for TLS cluster {}", info->name());
+      cluster_manager->thread_local_deferred_clusters_[info->name()] =
+          cluster_initialization_object;
+
+      // Invoke similar logic of onClusterAddOrUpdate.
+      ThreadLocalClusterCommand command = [&cluster_manager,
+                                           cluster_name = info->name()]() -> ThreadLocalCluster& {
+        auto* cluster_entry = cluster_manager->initializeClusterInlineIfExists(cluster_name);
+        ASSERT(cluster_entry != nullptr);
```

Comment:
Why is this invariant true?
Also if there are multiple `cluster_manager->update_callbacks_` (see code below), wouldn't each one of them call `initializeClusterInlineIfExists()` and multiple clusters will be created?

Reply:
great catch! fixed this; now deferred initialization should not fail if it occurs